content/posts/outbox.md

```diff
@@ -0,0 +1,74 @@
+# Outbox
+
+In the previous post we have shown how consistent messaging can be implemented by storing point-in-time state snapshots and using these snapshots for publishing outgoing messages. We discussed some pros and cons of this approach. This time we will focus on the alternative approach which is based on storing the outgoing messages before they are dispatched.
+
+Consistent messaging requires the ability to ensure exactly same side effects (in form of the outgoing messages) each time a copy of a given incoming message is processed. One way of fulfilling that requirement is to capture and persist these side effects as part of the state of the object. Once persisted, these side effects can be re-played each time a copy of the incoming message is received. The main advantage of this approach is simplicity. The user no longer needs to worry about making the processing logic deterministic (including accessing system clock or random number generator).
+
+Let's take a look at the code to understand better all the moving parts. First, we need to load the state of our business component. 
+
+```
+var stateContainer = await persister.LoadByCorrelationId(correlationId).ConfigureAwait(false) ?? new StateContainer { Id = correlationId };
+
+TransportOperation[] outgoingMessages;
+if (!stateContainer.OutboxState.ContainsKey(messageId))
+{
+```
+
+As you can see, the state container is extended with a data structure that holds the message processing state -- `OutboxState`. That structure is a map that associates ID of an incoming message with a collection of outgoing messages that were produced during processing. When a message handler receives a message, it loads the state and inspects the _Outbox_ to check if it contains an entry for the ID of the incoming message. If there is no corresponding entry, it means that a given message has not been processed yet.
+
+```
+var state = (T)stateContainer.State ?? new T();
+var (newState, pendingTransportOperations) = await handlerCallback(state, context);
+```
+
+In that case we invoke the business logic that returns a tuple containing a new value of the state and a collection of `pendingTransportOperations` -- messages that were generated and are to be dispatched. Now we need to take both these pieces, stick them into the container and make sure they are stored durably in a *single atomic operation*.
```

Comment:
The container here is a bit misleading (I though for a sec we are talking about DI).